src/gui/widgets/MicrotunerConfig.cpp

```diff
@@ -0,0 +1,647 @@
+ * \param index Index of the keymap to update; update all keymaps if -1 or out of range.
+ */
+void MicrotunerConfig::updateKeymapList(int index)
+{
+	if (index >= 0 && index < MaxKeymapCount)
+	{
+		m_keymapComboModel.replaceItem(index,
+			QString::number(index) + ": " + Engine::getSong()->getKeymap(index)->getDescription());
+	}
+	else
+	{
+		for (int i = 0; i < MaxKeymapCount; i++)
+		{
+			m_keymapComboModel.replaceItem(i,
+				QString::number(i) + ": " + Engine::getSong()->getKeymap(i)->getDescription());
+		}
+	}
+}
+
+
+/**
+ * \brief Fill all the scale-related values based on currently selected scale
+ */
+void MicrotunerConfig::updateScaleForm()
+{
+	Song *song = Engine::getSong();
+	if (song == nullptr) {return;}
+
+	auto newScale = song->getScale(m_scaleComboModel.value());
+
+	m_scaleNameEdit->setText(newScale->getDescription());
+
+	// fill in the intervals
+	m_scaleTextEdit->setPlainText("");
+	const std::vector<Interval> &intervals = newScale->getIntervals();
+	for (std::size_t i = 1; i < intervals.size(); i++)
+	{
+		m_scaleTextEdit->appendPlainText(intervals[i].getString());
+	}
+	// scroll back to the top
+	QTextCursor tmp = m_scaleTextEdit->textCursor();
+	tmp.movePosition(QTextCursor::Start);
+	m_scaleTextEdit->setTextCursor(tmp);
+}
+
+
+/**
+ * \brief Fill all the keymap-related values based on currently selected keymap
+ */
+void MicrotunerConfig::updateKeymapForm()
+{
+	Song *song = Engine::getSong();
+	if (song == nullptr) {return;}
+
+	auto newMap = song->getKeymap(m_keymapComboModel.value());
+
+	m_keymapNameEdit->setText(newMap->getDescription());
+
+	m_keymapTextEdit->setPlainText("");
+	const std::vector<int> &map = newMap->getMap();
+	for (std::size_t i = 0; i < map.size(); i++)
+	{
+		if (map[i] >= 0) {m_keymapTextEdit->appendPlainText(QString::number(map[i]));}
+		else {m_keymapTextEdit->appendPlainText("x");}
+	}
+	QTextCursor tmp = m_keymapTextEdit->textCursor();
+	tmp.movePosition(QTextCursor::Start);
+	m_keymapTextEdit->setTextCursor(tmp);
+
+	m_firstKeyModel.setValue(newMap->getFirstKey());
+	m_lastKeyModel.setValue(newMap->getLastKey());
+	m_middleKeyModel.setValue(newMap->getMiddleKey());
+	m_baseKeyModel.setValue(newMap->getBaseKey());
+	m_baseFreqModel.setValue(newMap->getBaseFreq());
+}
+
+
+/**
+ * \brief Validate the scale name and entered interval definitions
+ * \return true if input is valid, false if problems were detected
+ */
+bool MicrotunerConfig::validateScaleForm()
+{
+	auto fail = [=](QString message) {QMessageBox::critical(this, tr("Scale parsing error"), message);};
+
+	// check name
+	QString name = m_scaleNameEdit->text();
+	if (name.length() > 0 && name[0] == '!') {fail(tr("Scale name cannot start with an exclamation mark")); return false;}
+	if (name.contains('\n')) {fail(tr("Scale name cannot contain a new-line character")); return false;}
+
+	// check intervals
+	QStringList input = m_scaleTextEdit->toPlainText().split('\n', QString::SkipEmptyParts);
```

Comment:
`QString::SkipEmptyParts` is deprecated in Qt 5.15 in favor of `Qt::SkipEmptyParts` which was introduced in Qt 5.14. There are three more places where `QString::SkipEmptyParts` is used.